### PR TITLE
feat(Core/RealmList): port TrinityCore realm api

### LINKED ISSUE #6062: Problem when trying to turn on the emulator
**CURRENT BEHAVIOUR:**

For some unknown reason, when I added the boots library, the authserver did not work.

**EXPECTED BLIZZLIKE BEHAVIOUR:**

Before the addition of the library, you could compile the emulator without any problems.

**SOURCE:**

https://www.azerothcore.org/wiki/windows-requirements

**STEPS TO REPRODUCE THE PROBLEM:**

1. Clone or update the repository.
2. Attempt to run the authserver

**EXTRA NOTES:**

I think it could be the version of the library, so I have installed 1.70.0 and 174.0 but with both, I get the same result.

**AC HASH/COMMIT:**

https://github.com/azerothcore/azerothcore-wotlk/commit/78e1719c80e8c3174461dcee3a25bcd77438b450

**OPERATING SYSTEM:**

Windows 10

**MODULES:**

I have no modules.

**CUSTOMIZATIONS:**

A small modification in adding a custom race, but it is not the problem, because previously it could be compiled without problems.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98674062-problem-when-trying-to-turn-on-the-emulator?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- https://github.com/WarheadCore/Warhead/commit/2ea31027e5d5155fa88839aac38b175ec402e4ac
- https://github.com/WarheadCore/Warhead/commit/778f1725f8b085161d28a48806b1d2ec56a53e6b
- https://github.com/WarheadCore/Warhead/commit/b990a79fd8643159ebe5c302c564c9ef52b2b5c6
- Fix undefined behavior `ace reactor`

## Issues
- Closes #6062

## Tests Performed:
https://user-images.githubusercontent.com/18329778/119720234-ebc5a280-be93-11eb-9b02-239ef0c2f5c7.mp4

## How to Test the Changes:
- Just enter game

## Target Branch(es):
- [x] Master
 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: data/sql/updates/pending_db_auth/rev_1620114805872279900.sql [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  `build` int(11) NOT NULL,
+  `majorVersion` int(11) DEFAULT NULL,
+  `minorVersion` int(11) DEFAULT NULL,
+  `bugfixVersion` int(11) DEFAULT NULL,
+  `hotfixVersion` char(3) DEFAULT NULL,
+  `winAuthSeed` varchar(32) DEFAULT NULL,
+  `win64AuthSeed` varchar(32) DEFAULT NULL,
+  `mac64AuthSeed` varchar(32) DEFAULT NULL,
+  `winChecksumSeed` varchar(40) DEFAULT NULL,
+  `macChecksumSeed` varchar(40) DEFAULT NULL,
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_auth/rev_1620114805872279900.sql:17]:
```suggestion
  `majorVersion` INT DEFAULT NULL,
  `minorVersion` INT DEFAULT NULL,
  `bugfixVersion` INT DEFAULT NULL,
  `hotfixVersion` CHAR(3) DEFAULT NULL,
  `winAuthSeed` VARCHAR(32) DEFAULT NULL,
  `win64AuthSeed` VARCHAR(32) DEFAULT NULL,
  `mac64AuthSeed` VARCHAR(32) DEFAULT NULL,
  `winChecksumSeed` VARCHAR(40) DEFAULT NULL,
  `macChecksumSeed` VARCHAR(40) DEFAULT NULL,
```

Comment (Winfidonarleyan) [reply data/sql/updates/pending_db_auth/rev_1620114805872279900.sql]: Thanks

File: data/sql/updates/pending_db_auth/rev_1620114805872279900.sql [hunk 2]
```diff
@@ -0,0 +1,108 @@
+INSERT INTO `version_db_auth` (`sql_rev`) VALUES ('1620114805872279900');
+
+--
+-- Table structure for table `build_info`
+--
+DROP TABLE IF EXISTS `build_info`;
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_auth/rev_1620114805872279900.sql:6]: we need to update the wiki explaining what this table is

File: src/server/authserver/Server/AuthSocket.cpp [hunk 3]
```diff
@@ -386,6 +387,10 @@ bool AuthSocket::_HandleLogonChallenge()
         stmt = LoginDatabase.GetPreparedStatement(LOGIN_SEL_LOGONCHALLENGE);
         stmt->setString(0, _login);
 
+        _localizationName.resize(4);
+        for (int i = 0; i < 4; ++i)
+            _localizationName[i] = ch->country[4 - i - 1];
```
Comment (FrancescoBorzi) [inline src/server/authserver/Server/AuthSocket.cpp:392]: brackets

File: src/server/authserver/Server/AuthSocket.cpp [hunk 4]
```diff
@@ -464,62 +469,57 @@ bool AuthSocket::_HandleLogonChallenge()
...
-                    LOG_DEBUG("network", "'%s:%d' [AuthChallenge] account %s is using '%c%c%c%c' locale (%u)", socket().getRemoteAddress().c_str(), socket().getRemotePort(), _login.c_str (), ch->country[3], ch->country[2], ch->country[1], ch->country[0], GetLocaleByName(_localizationName) );
+                        LOG_DEBUG("network", "'%s:%d' [AuthChallenge] account %s is using '%c%c%c%c' locale (%u)", socket().getRemoteAddress().c_str(), socket().getRemotePort(), _login.c_str(), ch->country[3], ch->country[2], ch->country[1], ch->country[0], GetLocaleByName(_localizationName));
 #endif
-                    ///- All good, await client's proof
-                    _status = STATUS_LOGON_PROOF;
+                        ///- All good, await client's proof
+                        _status = STATUS_LOGON_PROOF;
+                    }
+                    else
+                        pkt << uint8(WOW_FAIL_VERSION_INVALID);
```
Comment (FrancescoBorzi) [inline src/server/authserver/Server/AuthSocket.cpp:522]: brackets

File: src/server/authserver/Server/AuthSocket.cpp [hunk 5]
```diff
@@ -831,20 +831,20 @@ ACE_INET_Addr const& AuthSocket::GetAddressForClient(Realm const& realm, ACE_INE
...
         // has all realms available in his local network
-        return realm.LocalAddress;
+        return *realm.LocalAddress;
     }
 
     // Check if connecting client is in the same network
-    if (IsIPAddrInNetwork(realm.LocalAddress, clientAddr, realm.LocalSubnetMask))
-        return realm.LocalAddress;
+    if (IsIPAddrInNetwork(*realm.LocalAddress, clientAddr, *realm.LocalSubnetMask))
+        return *realm.LocalAddress;
```
Comment (FrancescoBorzi) [inline src/server/authserver/Server/AuthSocket.cpp:844]: brackets

File: src/server/game/World/World.h [hunk 6]
```diff
@@ -334,8 +324,8 @@ class World: public IWorld
     void LoadWorldStates();
 
     /// Are we on a "Player versus Player" server?
-    bool IsPvPRealm() const { return (getIntConfig(CONFIG_GAME_TYPE) == REALM_TYPE_PVP || getIntConfig(CONFIG_GAME_TYPE) == REALM_TYPE_RPPVP || getIntConfig(CONFIG_GAME_TYPE) == REALM_TYPE_FFA_PVP); }
-    bool IsFFAPvPRealm() const { return getIntConfig(CONFIG_GAME_TYPE) == REALM_TYPE_FFA_PVP; }
+    bool IsPvPRealm() const;
+    bool IsFFAPvPRealm() const;
```
Comment (FrancescoBorzi) [inline src/server/game/World/World.h:328]: consider `[[nodiscard]]` whenever possible

File: src/server/shared/Realms/RealmList.cpp [hunk 7]
```diff
@@ -19,81 +22,188 @@ RealmList* RealmList::instance()
...
+            build.Build = fields[4].GetUInt32();
+            std::string windowsHash = fields[5].GetString();
+
+            if (windowsHash.length() == build.WindowsHash.size() * 2)
+                HexStrToByteArray(windowsHash, build.WindowsHash);
+
+            std::string macHash = fields[6].GetString();
+
+            if (macHash.length() == build.MacHash.size() * 2)
+                HexStrToByteArray(macHash, build.MacHash);
```
Comment (FrancescoBorzi) [inline src/server/shared/Realms/RealmList.cpp:61]: missing brackets everywhere

File: src/server/shared/Realms/RealmList.cpp [hunk 8]
```diff
@@ -19,81 +22,188 @@ RealmList* RealmList::instance()
...
 {
-    LOG_INFO("server", "Updating Realm List...");
+    LOG_DEBUG("server.authserver", "Updating Realm List...");
 
     PreparedStatement* stmt = LoginDatabase.GetPreparedStatement(LOGIN_SEL_REALMLIST);
     PreparedQueryResult result = LoginDatabase.Query(stmt);
 
+    std::map<RealmHandle, std::string> existingRealms;
+    for (auto const& [handle, realm] : _realms)
+        existingRealms[handle] = realm.Name;
```
Comment (FrancescoBorzi) [inline src/server/shared/Realms/RealmList.cpp:122]: same

File: src/server/shared/Realms/RealmList.cpp [hunk 9]
```diff
@@ -19,81 +22,188 @@ RealmList* RealmList::instance()
...
+            RealmHandle id{ realmId };
+
+            UpdateRealm(id, build, name, std::move(externalAddress.value()), std::move(localAddress.value()), std::move(localSubmask.value()), port, icon, flag,
+                timezone, (allowedSecurityLevel <= SEC_ADMINISTRATOR ? AccountTypes(allowedSecurityLevel) : SEC_ADMINISTRATOR), pop);
 
-            UpdateRealm(realmId, name, externalAddr, localAddr, submask, icon, flag, timezone, (allowedSecurityLevel <= SEC_ADMINISTRATOR ? AccountTypes(allowedSecurityLevel) : SEC_ADMINISTRATOR), pop, build);
+            if (!existingRealms.count(id))
+                LOG_INFO("server.authserver", "Added realm \"%s\" at %s:%u.", name.c_str(), externalAddressString.c_str(), port);
+            else
+                LOG_DEBUG("server.authserver", "Updating realm \"%s\" at %s:%u.", name.c_str(), externalAddressString.c_str(), port);
```
Comment (FrancescoBorzi) [inline src/server/shared/Realms/RealmList.cpp:186]: same

File: src/server/shared/Realms/RealmList.cpp [hunk 10]
```diff
@@ -19,81 +22,188 @@ RealmList* RealmList::instance()
...
+        return &itr->second;
+
+    return nullptr;
+}
+
+RealmBuildInfo const* RealmList::GetBuildInfo(uint32 build) const
+{
+    for (RealmBuildInfo const& clientBuild : _builds)
+        if (clientBuild.Build == build)
+            return &clientBuild;
```
Comment (FrancescoBorzi) [inline src/server/shared/Realms/RealmList.cpp:206]: same x2

File: src/server/worldserver/Master.cpp [hunk 11]
```diff
@@ -422,8 +427,52 @@ void Master::ClearOnlineAccounts()
...
 
     // Reset online status for all characters
     CharacterDatabase.DirectExecute("UPDATE characters SET online = 0 WHERE online <> 0");
 }
+
+bool LoadRealmInfo()
+{
+    QueryResult result = LoginDatabase.PQuery("SELECT id, name, address, localAddress, localSubnetMask, port, icon, flag, timezone, allowedSecurityLevel, population, gamebuild FROM realmlist WHERE id = %u", realm.Id.Realm);
+    if (!result)
+        return false;
```
Comment (FrancescoBorzi) [inline src/server/worldserver/Master.cpp:440]: same

**Comment by Winfidonarleyan** [comment]
Go to add more more brackets

File: data/sql/updates/pending_db_auth/rev_1620114805872279900.sql (hunk 2 above)
Comment (Winfidonarleyan) [reply data/sql/updates/pending_db_auth/rev_1620114805872279900.sql:6]: https://github.com/azerothcore/wiki/pull/466

File: data/sql/updates/pending_db_auth/rev_1620114805872279900.sql [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+CREATE TABLE `build_info` (
+  `build` int(11) NOT NULL,
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_auth/rev_1620114805872279900.sql:8]:
```suggestion
  `build` INT NOT NULL,
```

**Comment by Winfidonarleyan** [comment]
Wait merge #5640

File: src/server/shared/Realms/RealmList.cpp [hunk 13]
```diff
@@ -19,81 +22,210 @@ RealmList* RealmList::instance()
...
+                HexStrToByteArray(windowsHash, build.WindowsHash);
+            }
+
+            std::string macHash = fields[6].GetString();
+
+            if (macHash.length() == build.MacHash.size() * 2)
+            {
+                HexStrToByteArray(macHash, build.MacHash);
+            }
+
```
Comment (Kitzunu) [inline src/server/shared/Realms/RealmList.cpp:70]: unnecessary space

File: src/server/shared/Realms/Realm.h [hunk 14]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU AGPL v3 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-AGPL3
+ * Copyright (C) 2021+ WarheadCore <https://github.com/WarheadCore>
```
[the remark this answers is not in the document]
Comment (Winfidonarleyan) [reply src/server/shared/Realms/Realm.h:3]: Warhead - fork tc

File: data/sql/updates/pending_db_auth/rev_1620114805872279900.sql [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+  `mac64AuthSeed` VARCHAR(32) DEFAULT NULL,
+  `winChecksumSeed` VARCHAR(40) DEFAULT NULL,
+  `macChecksumSeed` VARCHAR(40) DEFAULT NULL,
+  PRIMARY KEY (`build`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+
+--
+-- Dumping data for table `build_info`
+--
+INSERT INTO `build_info` VALUES
```
[the remark this answers is not in the document]
Comment (Winfidonarleyan) [reply data/sql/updates/pending_db_auth/rev_1620114805872279900.sql:24]: Hm, no

**Comment by Winfidonarleyan** [comment]
Added video test PR

**Comment by pangolp** [comment]
Confirmed.

```
AzerothCore rev. 5a972f616762 2021-05-27 02:10:09 +0700 (pr-5626 branch) (Win64, Release) (authserver)
<Ctrl-C> to stop.

   █████╗ ███████╗███████╗██████╗  ██████╗ ████████╗██╗  ██╗
  ██╔══██╗╚══███╔╝██╔════╝██╔══██╗██╔═══██╗╚══██╔══╝██║  ██║
  ███████║  ███╔╝ █████╗  ██████╔╝██║   ██║   ██║   ███████║
  ██╔══██║ ███╔╝  ██╔══╝  ██╔══██╗██║   ██║   ██║   ██╔══██║
  ██║  ██║███████╗███████╗██║  ██║╚██████╔╝   ██║   ██║  ██║
  ╚═╝  ╚═╝╚══════╝╚══════╝╚═╝  ╚═╝ ╚═════╝    ╚═╝   ╚═╝  ╚═╝
                                 ██████╗ ██████╗ ██████╗ ███████╗
                                ██╔════╝██╔═══██╗██╔══██╗██╔════╝
                                ██║     ██║   ██║██████╔╝█████╗
                                ██║     ██║   ██║██╔══██╗██╔══╝
                                ╚██████╗╚██████╔╝██║  ██║███████╗
                                 ╚═════╝ ╚═════╝ ╚═╝  ╚═╝╚══════╝

     AzerothCore 3.3.5a  -  www.azerothcore.org

> Using configuration file       configs/authserver.conf.
> Using SSL version:             OpenSSL 1.0.2d 9 Jul 2015 (library: OpenSSL 1.0.2d 9 Jul 2015)
> Using Boost version:           1.74.0
> Using ACE version:             6.5.10

Max allowed open files is 1024
Opening DatabasePool 'acore_auth'. Asynchronous connections: 1, synchronous connections: 1.
MySQL client library: 5.7.33
MySQL server ver: 5.7.33-log 
Connected to MySQL database at 127.0.0.1
MySQL client library: 5.7.33
MySQL server ver: 5.7.33-log 
Connected to MySQL database at 127.0.0.1
DatabasePool 'acore_auth' opened successfully. 2 total connections running.
 
Started auth database connection pool.
Added realm "AzerothCore" at 127.0.0.1:8085.
Authserver listening to 0.0.0.0:3724
```

**Review by pangolp — APPROVED** [review]

**Comment by Winfidonarleyan** [comment]
Very good, thanks @pangolp

**Comment by tkn963** [comment]
Confirmed @ Ubuntu

**Review by Kitzunu — APPROVED** [review]
It fixes the serves for windows, so for me it is good

**Comment by Kitzunu** [comment]
Merged as it 
- fixes the Windows issue
- was confirmed working for Ubuntu
- was [confirmed working for Linux](https://discord.com/channels/217589275766685707/284406375495368704/847225645221019658)

**Comment by Kitzunu** [comment]
ty

**Comment by Winfidonarleyan** [comment]
Thanks all for tesing

**Pull request merged**